synapseclient/core/retry.py

```diff
@@ -181,6 +195,243 @@ def foo(a, b, c): return [a, b, c]
+    """
+    random_jitter = random.uniform(wait_random_lower, wait_random_upper)
+    time_to_wait = min(
+        (base_wait * (back_off_factor**retries)) + random_jitter,
+        max_back_off,
+    )
+    return time_to_wait
+
+
+async def with_retry_async(
+    function,
+    verbose: bool = False,
+    retry_status_codes: typing.List[int] = None,
+    expected_status_codes: typing.List[int] = None,
+    retry_errors: typing.List[str] = None,
+    retry_exceptions: typing.List[typing.Union[Exception, str]] = None,
+    retry_base_wait: float = DEFAULT_BASE_WAIT_ASYNC,
+    retry_wait_random_lower: float = DEFAULT_WAIT_RANDOM_LOWER_ASYNC,
+    retry_wait_random_upper: float = DEFAULT_WAIT_RANDOM_UPPER_ASYNC,
+    retry_back_off_factor: float = DEFAULT_BACK_OFF_FACTOR_ASYNC,
+    retry_max_back_off: float = DEFAULT_MAX_BACK_OFF_ASYNC,
+    retry_max_wait_before_failure: float = DEFAULT_MAX_WAIT_BEFORE_FAIL_ASYNC,
+) -> typing.Union[Exception, httpx.Response, typing.Any, None]:
+    """
+    Retries the given function under certain conditions. This is created such that it
+    will retry an unbounded number of times until the maximum wait time is reached. The
+    backoff is calculated using an exponential backoff algorithm with a random jitter.
+    The maximum backoff inbetween retries is capped at `retry_max_back_off`.
+
+    Arguments:
+        verbose: Whether to log debug messages
+        function: A function with no arguments. If arguments are needed, use a lambda
+            (see example).
+        retry_status_codes: What status codes to retry upon in the case of a
+            SynapseHTTPError.
+        expected_status_codes: If specified responses with any other status codes result
+            in a retry.
+        retry_errors: What reasons to retry upon, if
+            `function().response.json()['reason']` exists.
+        retry_exceptions: What types of exceptions, specified as strings or Exception
+            classes, to retry upon.
+        retry_base_wait: The base wait time inbetween retries.
+        retry_wait_random_lower: The lower bound of the random wait time.
+        retry_wait_random_upper: The upper bound of the random wait time.
+        retry_back_off_factor: The factor to increase the wait time by for each retry.
+        retry_max_back_off: The maximum wait time.
+        retry_max_wait_before_failure: The maximum wait time before failure.
+
+    Example: Using with_retry
+        Using ``with_retry_async`` to consolidate inputs into a list.
+
+            from synapseclient.core.retry import with_retry_async
+
+            async def foo(a, b, c): return [a, b, c]
+            result = await with_retry_async(lambda: foo("1", "2", "3"))
+    """
+    if not retry_status_codes:
+        retry_status_codes = [429, 500, 502, 503, 504]
+    if not expected_status_codes:
+        expected_status_codes = []
+    if not retry_errors:
+        retry_errors = []
+    if not retry_exceptions:
```

Comment:
nit: What's the advantage of doing this instead of setting default values for these args?

Reply:
https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Reply:
Good to know!